src/machine/uart_nxpmk66f18.go

```diff
@@ -0,0 +1,279 @@
+		arm.EnableIRQ(u.IRQNumber)
+	}
+}
+
+func (u UART) Disable() {
+	// adapted from Teensy core's serial_end
+
+	// check if the device has been enabled already
+	if !u.SCGC.HasBits(u.SCGCMask) {
+		return
+	}
+
+	u.Flush()
+
+	arm.DisableIRQ(u.IRQNumber)
+	u.C2.Set(0)
+
+	// reconfigure pin
+	u.RXPCR.Set(nxp.PORT_PCR0_PE | nxp.PORT_PCR0_PS | nxp.PORT_PCR0_MUX(1))
+	u.TXPCR.Set(nxp.PORT_PCR0_PE | nxp.PORT_PCR0_PS | nxp.PORT_PCR0_MUX(1))
+
+	// clear flags
+	u.S1.Get()
+	u.D.Get()
+	u.Buffer.Clear()
+}
+
+func (u UART) Flush() {
+	for u.Transmitting.Get() != 0 {
+		gosched()
+	}
+}
+
+// adapted from Teensy core's uart0_status_isr
+func (u UART) handleStatusInterrupt() {
+	// receive
+	if u.S1.HasBits(nxp.UART_S1_RDRF | nxp.UART_S1_IDLE) {
+		intrs := arm.DisableInterrupts()
+		avail := u.RCFIFO.Get()
+		if avail == 0 {
+			// The only way to clear the IDLE interrupt flag is
+			// to read the data register.  But reading with no
+			// data causes a FIFO underrun, which causes the
+			// FIFO to return corrupted data.  If anyone from
+			// Freescale reads this, what a poor design!  There
+			// write should be a write-1-to-clear for IDLE.
+			u.D.Get()
+			// flushing the fifo recovers from the underrun,
+			// but there's a possible race condition where a
+			// new character could be received between reading
+			// RCFIFO == 0 and flushing the FIFO.  To minimize
+			// the chance, interrupts are disabled so a higher
+			// priority interrupt (hopefully) doesn't delay.
+			// TODO: change this to disabling the IDLE interrupt
+			// which won't be simple, since we already manage
+			// which transmit interrupts are enabled.
+			u.CFIFO.Set(nxp.UART_CFIFO_RXFLUSH)
+			arm.EnableInterrupts(intrs)
+
+		} else {
+			arm.EnableInterrupts(intrs)
+
+			for {
+				u.Buffer.Put(u.D.Get())
+				avail--
+				if avail <= 0 {
+					break
+				}
+			}
+		}
+	}
+
+	// transmit
+	if u.C2.HasBits(nxp.UART_C2_TIE) && u.S1.HasBits(nxp.UART_S1_TDRE) {
+		data := make([]byte, 0, uartTXFIFODepth)
+		avail := uartTXFIFODepth - u.TCFIFO.Get()
+
+		// get avail bytes from ring buffer
+		for len(data) < int(avail) {
+			if b, ok := u.TXBuffer.Get(); ok {
+				data = append(data, b)
+			} else {
+				break
+			}
+		}
+
+		// write data to FIFO
+		l := len(data)
+		for i, b := range data {
+			if i == l-1 {
+				// only clear TDRE on last write, per the manual
+				u.S1.Get()
+			}
+			u.D.Set(b)
+		}
+
+		// if FIFO still has room, disable TIE, enable TCIE
+		if u.S1.HasBits(nxp.UART_S1_TDRE) {
+			u.C2.Set(uartC2TXCompleting)
+		}
+	}
+
+	// transmit complete
+	if u.C2.HasBits(nxp.UART_C2_TCIE) && u.S1.HasBits(nxp.UART_S1_TC) {
+		u.Transmitting.Set(0)
+		u.C2.Set(uartC2TXInactive)
+	}
+}
+
+// WriteByte writes a byte of data to the UART.
+func (u UART) WriteByte(c byte) error {
+	if !u.Configured {
+		return ErrNotConfigured
+	}
+
+	for !u.TXBuffer.Put(c) {
+		gosched()
+	}
+
+	u.Transmitting.Set(1)
+	u.C2.Set(uartC2TXActive)
+	return nil
+}
```

Comment:
You could replace this code (and all related code) with the few simple lines of code you have in the runtime `putchar` function. More complex buffers can always be added at a later time (also, you said this code doesn't actually work?).

Reply:
This code works now. I never really understood what Paul's code was doing exactly, so I took the general idea and reimplemented it.